x/oracle/keeper/prices.go

```diff
@@ -0,0 +1,189 @@
+		if b != nil {
+			//should alwyas be true since we don't delete prices from history round
+			k.cdc.MustUnmarshal(b, val.PriceList[i])
+			found = true
+		}
+	}
+
+	return
+}
+
+// RemovePrices removes a prices from the store
+func (k Keeper) RemovePrices(
+	ctx sdk.Context,
+	tokenId int32,
+
+) {
+	store := k.getPriceTRStore(ctx, tokenId)
+	//	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	iterator := store.Iterator(nil, nil)
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		store.Delete(iterator.Key())
+	}
+}
+
+// GetAllPrices returns all prices
+func (k Keeper) GetAllPrices(ctx sdk.Context) (list []types.Prices) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.PricesKeyPrefix))
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+
+	defer iterator.Close()
+
+	//	prevTokenId := uint32(0)
+	//	var val types.PriceWithTimeAndRound
+	var price types.Prices
+	prevTokenId := int32(0)
+	for ; iterator.Valid(); iterator.Next() {
+		tokenId, _, nextRoundId := parseKey(iterator.Key())
+		if prevTokenId == 0 {
+			prevTokenId = tokenId
+			price.TokenId = tokenId
+		} else if prevTokenId != tokenId && price.TokenId > 0 {
+			list = append(list, price)
+			prevTokenId = tokenId
+			price = types.Prices{TokenId: tokenId}
+		}
+		if nextRoundId {
+			price.NextRountId = binary.BigEndian.Uint64(iterator.Value())
+		} else {
+			var val types.PriceWithTimeAndRound
+			k.cdc.Unmarshal(iterator.Value(), &val)
+			price.PriceList = append(price.PriceList, &val)
+		}
+	}
+	if price.TokenId > 0 {
+		list = append(list, price)
+	}
+	return
+}
+
+func (k Keeper) AppendPriceTR(ctx sdk.Context, tokenId int32, priceTR types.PriceWithTimeAndRound) {
+	nextRoundId := k.GetNextRoundId(ctx, tokenId)
+	if nextRoundId != priceTR.RoundId {
+		return
+	}
+	store := k.getPriceTRStore(ctx, tokenId)
+	b := k.cdc.MustMarshal(&priceTR)
+	store.Set(types.PricesRoundKey(nextRoundId), b)
+	k.IncreaseNextRoundId(ctx, tokenId)
+}
+
+//func(k Keeper) SetPriceTR(ctx sdk.Context, tokenId int32, priceTR){}
+
+func (k Keeper) GetPriceTRRoundId(ctx sdk.Context, tokenId int32, roundId uint64) (price types.PriceWithTimeAndRound, found bool) {
+	store := k.getPriceTRStore(ctx, tokenId)
+
+	b := store.Get(types.PricesRoundKey(roundId))
+	if b == nil {
+		return
+	}
+
+	k.cdc.Unmarshal(b, &price)
+	found = true
+	return
+}
+
+func (k Keeper) GetPriceTRLatest(ctx sdk.Context, tokenId int32) (price types.PriceWithTimeAndRound, found bool) {
+	//	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.PricesKeyPrefix))
+	//	store = prefix.NewStore(store, types.PricesKey(tokenId))
+	store := k.getPriceTRStore(ctx, tokenId)
+	nextRoundIdB := store.Get(types.PricesNextRountIdKey)
+	if nextRoundIdB == nil {
+		return
+	}
+	nextRoundId := binary.BigEndian.Uint64(nextRoundIdB)
+	b := store.Get(types.PricesRoundKey(nextRoundId - 1))
+	if b != nil {
+		//should always be true
+		k.cdc.Unmarshal(b, &price)
```

Comment:
k.cdc.Unmarshal has error return which need to be handle. can k.cdc.MustMarshal ?